File: 2021/Day14.py
def main():
    with open("input.txt") as f:
        input = f.read()
        print(f"Part 1 Answer: {part1(input)}")
        print(f"Part 2 Answer: {part2(input)}")

def part1(input):
    data = input.split("\n")
    polymer = data[0]
    rules = {d.split(" -> ")[0] : d.split(" -> ")[1] for d in data[2:]}
    for _ in range(10):
        index = 0
        while index < len(polymer)-1:
            pair = polymer[index] + polymer[index + 1]
            if pair in rules:
                polymer = polymer[:index+1] + rules[pair] + polymer[index+1:]
                index += 1
            index += 1
    frequencies = {}
    for char in polymer:
        if char in frequencies:
            frequencies[char] += 1
        else:
            frequencies[char] = 1
    quantities = list(frequencies.values())
    return max(quantities) - min(quantities)

def part2(input):
    data = input.split("\n")
    pairs = {data[0][i] + data[0][i + 1] : data[0].count(data[0][i] + data[0][i + 1]) for i in range(len(data[0])-1)}
    rules = {d.split(" -> ")[0] : d.split(" -> ")[1] for d in data[2:]}
    for _ in range(40):
        elemPairs = list(pairs.keys())
        freqs = list(pairs.values())
        for i in range(len(elemPairs)):
            if elemPairs[i] in rules:
                insertion = rules[elemPairs[i]]
                if elemPairs[i][0] + insertion in pairs:
                    pairs[elemPairs[i][0] + insertion] += freqs[i]
                else:
                    pairs[elemPairs[i][0] + insertion] = freqs[i]
                if insertion + elemPairs[i][1] in pairs:
                    pairs[insertion + elemPairs[i][1]] += freqs[i]
                else:
                    pairs[insertion + elemPairs[i][1]] = freqs[i]
                pairs[elemPairs[i]] -= freqs[i]
    frequencies = {} 
    for pair in pairs:
        if pair[1] in frequencies:
            frequencies[pair[1]] += pairs[pair]
        else:
            frequencies[pair[1]] = pairs[pair]
    quantities = list(frequencies.values())
    return max(quantities) - min(quantities)

if __name__ == '__main__':
    main()

File: 2023/Day14.cs
class Day14
{
    static void Main()
    {
        string[] data = File.ReadLines("input.txt").ToArray();
        char[,] grid = new char[data.Length, data[0].Length];
        for (int i = 0; i < data.Length; i++)
        {
            for (int j = 0; j < data.Length; j++)
            {
                grid[i, j] = data[i][j];
            }
        }
        Console.WriteLine("Part 1 Answer: " + Part1(grid));
        Console.WriteLine("Part 2 Answer: " + Part2(grid));
    }

    static int Part1(char[,] grid)
    {
        int total = 0;
        for (int i = 0; i < grid.GetLength(1); i++)
        {
            int lastStop = -1;
            for (int j = 0; j < grid.GetLength(0); j++)
            {
                if (grid[j, i] == '#')
                {
                    lastStop = j;
                }
                else if (grid[j, i] == 'O')
                {
                    total += grid.GetLength(0) - (lastStop + 1);
                    lastStop++;
                }
            }
        }
        return total;
    }

    static int Part2(char[,] grid)
    {
        List<char[,]> history = new List<char[,]>();
        char[,] copy = new char[grid.GetLength(0), grid.GetLength(1)];
        Array.Copy(grid, copy, grid.Length);
        history.Add(copy);
        for (int n = 0; n < 1000000000; n++)
        {
            for (int i = 0; i < grid.GetLength(1); i++)
            {
                int lastStop = -1;
                for (int j = 0; j < grid.GetLength(0); j++)
                {
                    if (grid[j, i] == '#')
                    {
                        lastStop = j;
                    }
                    else if (grid[j, i] == 'O')
                    {
                        grid[j, i] = '.';
                        grid[lastStop + 1, i] = 'O';
                        lastStop++;
                    }
                }
            }
            for (int i = 0; i < grid.GetLength(0); i++)
            {
                int lastStop = -1;
                for (int j = 0; j < grid.GetLength(1); j++)
                {
                    if (grid[i, j] == '#')
                    {
                        lastStop = j;
                    }
                    else if (grid[i, j] == 'O')
                    {
                        grid[i, j] = '.';
                        grid[i, lastStop + 1] = 'O';
                        lastStop++;
                    }
                }
            }
            for (int i = 0; i < grid.GetLength(1); i++)
            {
                int lastStop = grid.GetLength(0);
                for (int j = grid.GetLength(0) - 1; j >= 0; j--)
                {
                    if (grid[j, i] == '#')
                    {
                        lastStop = j;
                    }
                    else if (grid[j, i] == 'O')
                    {
                        grid[j, i] = '.';
                        grid[lastStop - 1, i] = 'O';
                        lastStop--;
                    }
                }
            }
            for (int i = 0; i < grid.GetLength(0); i++)
            {
                int lastStop = grid.GetLength(1);
                for (int j = grid.GetLength(1) - 1; j >= 0; j--)
                {
                    if (grid[i, j] == '#')
                    {
                        lastStop = j;
                    }
                    else if (grid[i, j] == 'O')
                    {
                        grid[i, j] = '.';
                        grid[i, lastStop - 1] = 'O';
                        lastStop--;
                    }
                }
            }
            for (int k = 0; k < history.Count; k++)
            {
                bool match = true;
                for (int i = 0; i < grid.GetLength(0); i++)
                {
                    for (int j = 0; j < grid.GetLength(1); j++)
                    {
                        if (grid[i, j] != history[k][i, j])
                        {
                            match = false;
                            break;
                        }
                    }
                }
                if (match)
                {
                    int index = k + (1000000000 - k) % (n + 1 - k);
                    int total = 0;
                    for (int i = 0; i < grid.GetLength(0); i++)
                    {
                        for (int j = 0; j < grid.GetLength(1); j++)
                        {
                            if (history[index][i, j] == 'O')
                            {
                                total += grid.GetLength(0) - i;
                            }
                        }
                    }
                    return total;
                }
            }
            copy = new char[grid.GetLength(0), grid.GetLength(1)];
            Array.Copy(grid, copy, grid.Length);
            history.Add(copy);
        }
        return -1;
    }
}
